### ISSUE TITLE
[BUG] Too many connections while using AWS RDS as backend

### ISSUE THREAD

**Issue by murphycj** [mlflow/mlflow#3203]
### Willingness to contribute
- [ ] Yes. I can contribute a fix for this bug independently.
- [X] Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.
- [ ] No. I cannot contribute a bug fix at this time.

### System information
- **Have I written custom code (as opposed to using a stock example script provided in MLflow)**: 
No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Linux 4.14.181-142.260.amzn2.x86_64
- **MLflow installed from (source or binary)**:
pip install mlflow
- **MLflow version (run ``mlflow --version``)**:
1.10.0
- **Python version**:
3.7.7
- **npm version, if running the dev UI**:
NA
- **Exact command to reproduce**:

### Describe the problem
Apologies if this is a duplicate. I did see a similar issue (#1565), but the solution is unclear. I also see this issue #2650 but I am not sure if it is the same as mine. So I though I'd post my issue.

I am using AWS-RDS (postgresql) as a backend for mlflow while training with tensorflow/keras on EC2 instances. Also note that I am using talos ([github](https://github.com/autonomio/talos)) for parameter optimization. Everything runs fine with just one worker running, but if I have 2+ workers running on separate EC2 instances then I get this error indicating the RDS database has too many connections. See the error message produce by mlflow below.

<details><summary>Click for error message</summary>
<p>

```
Traceback (most recent call last):
  File "/home/ec2-user/anaconda3/envs/tensorflow2_latest_p37/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 2280, in _wrap_pool_connect
    e, dialect, self
  File "/home/ec2-user/anaconda3/envs/tensorflow2_latest_p37/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1547, in _handle_dbapi_exception_noconnection
    util.raise_from_cause(sqlalchemy_exception, exc_info)
  File "/home/ec2-user/anaconda3/envs/tensorflow2_latest_p37/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 398, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/home/ec2-user/anaconda3/envs/tensorflow2_latest_p37/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 152, in reraise
    raise value.with_traceback(tb)
  File "/home/ec2-user/anaconda3/envs/tensorflow2_latest_p37/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 2276, in _wrap_pool_connect
    return fn()
  File "/home/ec2-user/anaconda3/envs/tensorflow2_latest_p37/lib/python3.7/site-packages/sqlalchemy/pool/base.py", line 303, in unique_connection
    return _ConnectionFairy._checkout(self)
  File "/home/ec2-user/anaconda3/envs/tensorflow2_latest_p37/lib/python3.7/site-packages/sqlalchemy/pool/base.py", line 773, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "/home/ec2-user/anaconda3/envs/tensorflow2_latest_p37/lib/python3.7/site-packages/sqlalchemy/pool/base.py", line 492, in checkout
    rec = pool._do_get()
  File "/home/ec2-user/anaconda3/envs/tensorflow2_latest_p37/lib/python3.7/site-packages/sqlalchemy/pool/impl.py", line 139, in _do_get
    self._dec_overflow()
  File "/home/ec2-user/anaconda3/envs/tensorflow2_latest_p37/lib/python3.7/site-packages/sqlalchemy/util/langhelpers.py", line 68, in __exit__
    compat.reraise(exc_type, exc_value, exc_tb)
  File "/home/ec2-user/anaconda3/envs/tensorflow2_latest_p37/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 153, in reraise
    raise value
  File "/home/ec2-user/anaconda3/envs/tensorflow2_latest_p37/lib/python3.7/site-packages/sqlalchemy/pool/impl.py", line 136, in _do_get
    return self._create_connection()
  File "/home/ec2-user/anaconda3/envs/tensorflow2_latest_p37/lib/python3.7/site-packages/sqlalchemy/pool/base.py", line 308, in _create_connection
    return _ConnectionRecord(self)
  File "/home/ec2-user/anaconda3/envs/tensorflow2_latest_p37/lib/python3.7/site-packages/sqlalchemy/pool/base.py", line 437, in __init__
    self.__connect(first_connect_check=True)
  File "/home/ec2-user/anaconda3/envs/tensorflow2_latest_p37/lib/python3.7/site-packages/sqlalchemy/pool/base.py", line 652, in __connect
    connection = pool._invoke_creator(self)
  File "/home/ec2-user/anaconda3/envs/tensorflow2_latest_p37/lib/python3.7/site-packages/sqlalchemy/engine/strategies.py", line 114, in connect
    return dialect.connect(*cargs, **cparams)
  File "/home/ec2-user/anaconda3/envs/tensorflow2_latest_p37/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 489, in connect
    return self.dbapi.connect(*cargs, **cparams)
  File "/home/ec2-user/anaconda3/envs/tensorflow2_latest_p37/lib/python3.7/site-packages/psycopg2/__init__.py", line 130, in connect
    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) FATAL:  remaining connection slots are reserved for non-replication superuser connections
```

</p>
</details>

I also attached a screenshot from AWS RDS dashboard showing the number of connections overtime (note the screenshot was when I had only one mlflow worker running). I would expect there to just be two connections: one from EC2 worker and one from mlflow UI server, but I instead see 50+.
<img width="961" alt="Screen Shot 2020-07-30 at 8 35 00 PM" src="https://user-images.githubusercontent.com/3837196/88987580-6b023e00-d2a4-11ea-850d-6bdc03426148.png">

### Code to reproduce issue
Here is a distilled down version of the code I am running to give you an idea of how I am doing this:

```python
import json

from sklearn.model_selection import train_test_split
import tensorflow as tf
import mlflow
from mlflow import tensorflow as mlt
import talos

tf.config.threading.set_intra_op_parallelism_threads(8)
tf.config.threading.set_inter_op_parallelism_threads(8)
AUTOTUNE = tf.data.experimental.AUTOTUNE

config = json.load(open('config.json'))

mlflow.set_tracking_uri(
    'postgresql://{}:{}@{}:{}/mlflow'.format(
        config['db']['USER'],
        config['db']['PASSWORD'],
        config['db']['HOST'],
        config['db']['PORT']
    ))
mlt.autolog(1)

def model_scan(x_train, y_train, x_val, y_val, params):

    with mlflow.start_run(experiment_id='9'):

        for name, val in params.items():
            mlflow.log_param(name, val)

        dataset = load_data(x_train, y_train, params)
        dataset_test = load_data(x_val, y_val, params)

        dataset = dataset.prefetch(buffer_size=AUTOTUNE)
        dataset_test = dataset_test.prefetch(buffer_size=AUTOTUNE)

        model = model1(params)
        model_history = model.fit(
            dataset.repeat().batch(params['batch_size']),
            epochs=params['epochs'],
            steps_per_epoch=params['steps_per_epoch'],
            validation_data=dataset_val.batch(params['batch_size']),
            validation_freq=params['validation_freq'],
            callbacks=[
                MetricCallback(),
                talos.utils.early_stopper(params['epochs'])
                ])

    return model_history, model

x_train, x_val, y_train, y_val = train_test_split(X, y, test_size=0.1)

scan_object = talos.Scan(
    x=x_train,
    y=y_train,
    model=model_scan,
    params=params,
    fraction_limit=0.1,
    x_val=x_val,
    y_val=y_val,
    experiment_name='',
    random_method='quantum')
```

### What component(s), interfaces, languages, and integrations does this bug affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [X] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

**Comment by dmatrix** [comment]
@murphycj thanks for filing this issue and detailed explanation and reproducible code.

Can you try using [mlflow.log_batch()](https://www.mlflow.org/docs/latest/python_api/mlflow.tracking.html#mlflow.tracking.MlflowClient.log_batch)? to see if that helps mitigate?
Your changed code would like as follows:

```
with mlflow.start_run(experiment_id='9') as run:
    run_id = run.info.run_uuid
   ...
   mlflow.log_batch(...)
```

**Comment by murphycj** [comment]
@dmatrix Thanks for your reply. I did made your suggested modifications (see the code below), but I still get the same issue.

I should also add I do have a MetricCallback class defined that gets called during training and I do have metrics being logged to mlflow. I thought that that might be part of the issue, but when I removed the callback it made no apparent difference.


```python
import json

from sklearn.model_selection import train_test_split
import tensorflow as tf
import mlflow
from mlflow import tensorflow as mlt
from mlflow.entities import Metric, Param
from mlflow.tracking import MlflowClient
import talos

tf.config.threading.set_intra_op_parallelism_threads(8)
tf.config.threading.set_inter_op_parallelism_threads(8)
AUTOTUNE = tf.data.experimental.AUTOTUNE

config = json.load(open('config.json'))

mlflow.set_tracking_uri(
    'postgresql://{}:{}@{}:{}/mlflow'.format(
        config['db']['USER'],
        config['db']['PASSWORD'],
        config['db']['HOST'],
        config['db']['PORT']
    ))
mlt.autolog(1)

class MetricCallback(tf.keras.callbacks.Callback):

    def __init__(self, mlflow_client, run_id):
        self.mlflow_client=mlflow_client
        self.run_id=run_id

    def on_epoch_end(self, epoch, logs=None):

        if not logs:
            return

        metrics = []

        if logs.get('accuracy'):
            metrics.append(
                Metric(
                    key='accuracy',
                    value=float(logs['accuracy']),
                    timestamp=0,
                    step=epoch
                    )
            )

        if logs.get('val_loss'):
            metrics.append(
                Metric(
                    key='val_loss',
                    value=float(logs['val_loss']),
                    timestamp=0,
                    step=epoch
                    )
            )

        if logs.get('val_accuracy'):
            metrics.append(
                Metric(
                    key='val_accuracy',
                    value=float(logs['val_accuracy']),
                    timestamp=0,
                    step=epoch
                    )
            )

        if metrics:
            self.mlflow_client.log_batch(self.run_id, metrics=metrics)


def model_scan(x_train, y_train, x_val, y_val, params):

    with mlflow.start_run(experiment_id='9', run_name='v3') as run:
        mlflow_client = MlflowClient()
        run_id = run.info.run_uuid

        params_list = [Param(i[0], str(i[1])) for i in params.items()]
        mlflow_client.log_batch(run_id, params=params_list)

        dataset = load_data(x_train, y_train, params)
        dataset_test = load_data(x_val, y_val, params)

        dataset = dataset.prefetch(buffer_size=AUTOTUNE)
        dataset_test = dataset_test.prefetch(buffer_size=AUTOTUNE)

        print('fitting')
        with tf.device('/CPU:0'):
            model = model1(params)
            model_history = model.fit(
                dataset.repeat().batch(params['batch_size']),
                epochs=params['epochs'],
                steps_per_epoch=params['steps_per_epoch'],
                validation_data=dataset_val.batch(params['batch_size']),
                validation_freq=params['validation_freq'],
                callbacks=[
                    MetricCallback(mlflow_client=mlflow_client,run_id=run_id),
                    talos.utils.early_stopper(params['epochs'])])

        metrics = mlfow_finish(run_id, model, dataset_test, dataset_val, params, data)

        mlflow_client.log_batch(run_id, metrics=metrics)

    return model_history, model

x_train, x_val, y_train, y_val = train_test_split(X, y, test_size=0.1)

scan_object = talos.Scan(
    x=x_train,
    y=y_train,
    model=model_scan,
    params=params,
    fraction_limit=0.1,
    x_val=x_val,
    y_val=y_val,
    experiment_name='',
    random_method='quantum')

```

**Comment by dmatrix** [comment]
I suspect the issue might be your [limited default connection pool ](https://stackoverflow.com/questions/30778015/how-to-increase-the-max-connections-in-postgres)is exhausted. A similar issue wrt to [SQLALchemy DB](https://github.com/mlflow/mlflow/issues/1452) was resolved. 

Do you want to try[ expanding connections pool](https://mlflow.org/docs/latest/tracking.html#sqlalchemy-options) and setting these environment variables?

**Comment by murphycj** [comment]
@dmatrix Sorry for belated reply, just been busy. I tried your new suggestion via the following:

```
export MLFLOW_SQLALCHEMYSTORE_POOL_SIZE=1
export MLFLOW_SQLALCHEMYSTORE_MAX_OVERFLOW=1
```

Mlflow does seem to recognize that I set those options because before each epoch it prints out this message:

```
2020/08/20 01:00:58 INFO mlflow.store.db.utils: Create SQLAlchemy engine with pool options {'pool_size': 1, 'max_overflow': 1}
```

However, that still does not seem to work. I am still getting the number of connections going up to ~15-20. It seems that it creates a new connection for each epoch, but after a dozen or two epochs it resets the number of connections.
